Hello Tzvi!

My name is Olga. I'm happy to review your project today. 

When I will see mistake at the first time, I will just point it out. I let you find it and fix it by yourself. I'm trying to prepare you to work as an Data Analyst. SO, at a real job, your team lead will do the same. But if you can't handle this task yet, I will give you a more accurate hint at the next iteration.

You will find my comments below - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Success. Everything is done succesfully.
</div>


<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Remarks. Some recommendations.
</div>


<div class="alert alert-block alert-danger">
    
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
    
Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>


You can answer me by using this: 

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Overallv reviewer's comment</b> <a class="tocSkip"></a>
    
Tzvi, thank you for sending your project. 

You have provided a good job! You correctly used subqueries to solve the tasks. 

<span class="text-danger">However, one query should be improved. Please check my comments and I believe you can easily fix them!</span>
    
Keep up the good work, and come back soon!
    
</div>

# Book lover

I've been given a database of one of the apps for book lovers services. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product. 

For this information to be used effectively, I have to:
1. Find the number of books released after January 1, 2000.
1. Find the number of user reviews and the average rating for each book.
1. Identify the publisher that has released the greatest number of books with more than 50 pages
1. Identify the author with the highest average book rating (look only at books with at least 50 ratings).
1. Find the average number of text reviews among users who rated more than 50 books.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Excellent introduction! Let's start!
</div>

In [1]:
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Yep! All is right here!</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
It would be more informative to show tables and check missing values and duplicates. </div>

# number of books released

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Could you add conclusions for each step?</div>

In [2]:
qrl = "SELECT * FROM books ORDER BY publication_date DESC LIMIT 10;"

pd.io.sql.read_sql(qrl, con = engine)

book_id  author_id                                              title  \
0       43        377                         A Quick Bite (Argeneau #1)   
1      635        166                                  The Art of Loving   
2      445        612                                            Monster   
3      293         80                                         Ham on Rye   
4      993         80                                              Women   
5      414        626                                            Macbeth   
6      361        240  J.R.R. Tolkien 4-Book Boxed Set: The Hobbit an...   
7       36        169                                   A Moveable Feast   
8      428        301            Men Are from Mars  Women Are from Venus   
9       87        626                                     As You Like It   

   num_pages publication_date  publisher_id  
0        360       2020-03-31            28  
1        192       2019-08-06           130  
2        281       2019-03-05            14  
3        288       2014-07-29            99  
4        291       2014-07-29            99  
5        249       2013-07-01           268  
6       1728       2012-09-25            33  
7        192       2012-09-06           309  
8        368       2012-04-03           128  
9        263       2011-08-23           275

In [3]:
pop = "SELECT * FROM books ORDER BY publication_date LIMIT 10;"

pd.io.sql.read_sql(pop, con = engine)

book_id  author_id                                              title  \
0       49        569                           A Streetcar Named Desire   
1      823         20  The Philosophy of Andy Warhol (From A to B and...   
2      869        542                                        The Shining   
3      321        197                                Homage to Catalonia   
4      126        173                               Brideshead Revisited   
5      153        324  Cloudy With a Chance of Meatballs (Cloudy with...   
6      775        253  The Last of the Mohicans (The Leatherstocking ...   
7      165        544                                          Creepshow   
8      752        307                            The Hotel New Hampshire   
9      213        354                      Emily of New Moon (Emily  #1)   

   num_pages publication_date  publisher_id  
0        107       1952-12-01            97  
1        272       1977-04-06           178  
2        659       1980-07-01           190  
3        232       1980-10-22           124  
4        351       1982-01-30            31  
5         32       1982-04-01            25  
6        410       1982-06-01            38  
7         66       1982-08-01           226  
8        520       1982-10-22            51  
9        339       1983-04-01            81

In [10]:
book_num = '''SELECT COUNT(book_id) 
              FROM books 
              WHERE publication_date > '1.1.2000';'''

pd.io.sql.read_sql(book_num, con = engine)

count
0    819

I found 819 books after 1.1.2000. Doesn't seem like to much to be honest, the app needs to include some new books.

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
We need books after January 1, 2000. We do not need other dates.</div>

# Number of user reviews and the average rating for each book

In [19]:
num_rev_avg_rat = '''WITH

                     re AS (SELECT COUNT(review_id) AS number_of_reviews,
                                   book_id
                            FROM reviews
                            GROUP BY book_id),
                            
                     ra AS (SELECT AVG(rating) AS avg_rating,
                                   book_id
                             FROM ratings
                             GROUP BY book_id)
                    
                     
                     SELECT re.number_of_reviews, 
                            ra.avg_rating,
                            b.title
                     FROM books AS b
                     LEFT OUTER JOIN ra ON b.book_id = ra.book_id
                     LEFT OUTER JOIN re ON b.book_id = re.book_id;'''

pd.io.sql.read_sql(num_rev_avg_rat, con = engine)

number_of_reviews  avg_rating  \
0                  2.0    4.500000   
1                  2.0    4.500000   
2                  5.0    4.250000   
3                  2.0    4.000000   
4                  4.0    4.285714   
..                 ...         ...   
995                NaN    5.000000   
996                NaN    3.666667   
997                NaN    4.000000   
998                NaN    4.000000   
999                NaN    3.000000   

                                                 title  
0            The Body in the Library (Miss Marple  #3)  
1                                            Galápagos  
2                             A Tree Grows in Brooklyn  
3    Undaunted Courage: The Pioneering First Missio...  
4                                          The Prophet  
..                                                 ...  
995   The Cat in the Hat and Other Dr. Seuss Favorites  
996    Anne Rice's The Vampire Lestat: A Graphic Novel  
997                          Essential Tales and Poems  
998                               Leonardo's Notebooks  
999                            The Natural Way to Draw  

[1000 rows x 3 columns]

In [20]:
pd.io.sql.read_sql(num_rev_avg_rat, con = engine).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   number_of_reviews  994 non-null    float64
 1   avg_rating         1000 non-null   float64
 2   title              1000 non-null   object 
dtypes: float64(2), object(1)
memory usage: 23.6+ KB


It does not seem like a lot of people are engaging in reviews, even if the book has a very high rating. 

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
We need more simply query here. Main talbe is books. And we need 2 subqueries: for count and for avg.

</div>

# The publisher that has released the greatest number of books

In [6]:
publish_master = '''WITH
                          
                    p AS (SELECT publisher,
                                 publisher_id
                          FROM publishers),
                    b AS (SELECT publisher_id,
                                 COUNT(book_id) AS num_books
                          FROM books
                          WHERE num_pages > 50
                          GROUP BY publisher_id)
                          
                    SELECT b.num_books,
                           p.publisher
                    FROM b
                    LEFT OUTER JOIN p ON b.publisher_id = p.publisher_id
                    ORDER BY b.num_books DESC LIMIT 1;'''

pd.io.sql.read_sql(publish_master, con = engine)

num_books      publisher
0         42  Penguin Books

I found the most popular publisher, bot it does not seem like a big amount of books, maybe the app in general needs a larger amount of books.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Yes, you are right!</div>

# Author with the highest average book rating

In [7]:
great_auther = '''WITH

                  br AS (SELECT COUNT(r.rating) AS num_rating,
                                b.author_id
                         FROM books as b
                         LEFT OUTER JOIN ratings AS r ON b.book_id = r.book_id 
                         GROUP BY b.author_id)
                         
                  SELECT AVG(ra.rating) AS avg_rating,
                         a.author
                  FROM br
                  LEFT OUTER JOIN books AS bo ON br.author_id = bo.author_id
                  LEFT OUTER JOIN ratings AS ra ON bo.book_id = ra.book_id
                  LEFT OUTER JOIN authors AS a ON br.author_id = a.author_id
                  WHERE br.num_rating > 50
                  GROUP BY a.author
                  ORDER BY avg_rating DESC LIMIT 1;'''

pd.io.sql.read_sql(great_auther, con = engine)

avg_rating                      author
0    4.288462  J.K. Rowling/Mary GrandPré

Not surprising to say the least.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Yes, Harry Potter is popular. :) </div>

# Average number of text reviews

In [9]:
soso = '''SELECT AVG(final.review_count)
          FROM (SELECT re.username,
                       tr.rating_count,
                       COUNT(re.review_id) AS review_count
                FROM (SELECT username,
                             COUNT(rating) as rating_count
                      FROM ratings
                      GROUP BY username
                      HAVING COUNT(rating) > 50) AS tr
                LEFT JOIN reviews AS re ON re.username = tr.username
                GROUP BY re.username, tr.rating_count) AS final;'''
pd.io.sql.read_sql(soso, con = engine)

avg
0  24.333333

Even users that use the app constantly, don't like to use the review platform.

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

We need  number of text reviews among users who rated **more than 50 books**. Not included.

</div>

# Conclusion

1. In the first task we sew that the number of books, since the beginning of the 2000's, in the app is 821.  
   Which doesn't seem like much, the app needs to upload more books from the last two decades.
1. From the second task we sew that the average rating does seem average,  
   but the app users don't have a lot to go on with such a small number of reviews for each book.
1. From the third task, we see that Penguin Books is a really big name in publishing. 
1. It would have bean surprising if J.K. Rowling wasn't the most top rated author. I mean, with a fan base like hers.
1. Seems like only half of all rating of devoted users come along with a review.  
   Maybe the app needs to improve the review platform.  

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Thank you for sup up whole project here! </div>